go/v1/statement.go

```diff
@@ -0,0 +1,65 @@
+/*
+ * Wrapper APIs for in-toto attestation Statement layer protos.
+*/
+
+package v1
+
+import (
+	"errors"
+	
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const StatementTypeUri = "https://in-toto.io/Statement/v1"
+
+func NewStatementPb(subject []*ResourceDescriptor,
```

Comment:
Thanks @marcelamelara for working on making our life (library consumers) easier, I really appreciate it :)

My two cents, but please take them with a grain of salt, I do not have all the context of your vision on how these wrappers might evolve over time.

That said, I have mixed feelings about the usefulness of these "constructors". Apart of the validation part, which I like, creating this additional piece of indirection might result in some maintenance burden. i.e if the statement schema changes, modifications on the function signature might be required.

I'd argue too that this function doesn't simplify much the life of a `caller`, In the end is limiting itself to assemble in this case a statement directly from the function args, and `structpb.Struct`s are quite code intensive to assemble in my opinion. 

What I really like though is the validation logic, and this is something I'd like to use from Chainloop itself.

So I am wondering if as an alternative, we could just extend the generated proto message with a `validation function`. That way, I, as a consumer of these proto bindings, could decide how to assemble the `ResourceDescriptor` or `Statement` while still be able to leverage this validation logic.  

FWIW; I do something similar [here](https://github.com/chainloop-dev/chainloop/blob/ac7d2f5df369a6000185bf17a49f86557b6c8b1c/app/controlplane/api/workflowcontract/v1/crafting_schema_validations.go#L24). Where a protoc generated message (`v1.CraftingSchema` ) is "extended".

In any case, in the current state, this PR LGTM!

Thank you!

Reply:
@migmartri Thank you for your comments! I must admit I'm not a very proficient Go programmer, so some of the oddities in my implementation might stem from that.

For simple Statements/RDs, I think the constructors can be convenient shorthand; cases needing more customizability can still always build the messages directly (as we do in many of the tests). I think the issue of added fields in messages requiring changes to the function signature is much more likely to occur for predicates, as we don't expect the Statement layer to change very often, especially within the same major version.

That said, we (the attestation maintainers) have decided it may be best for the in-toto-golang implementation to host any wrapper APIs, so we will leave that aspect up to the maintainers there. I really like your suggestion to extend the pre-generated message (I tried to find whether this is possible, and couldn't find any info), so I will likely implement your suggestion instead.

> structpb.Structs are quite code intensive to assemble in my opinion.

Don't disagree... what do you suggest as an alternative? This seems like a general problem with using structpb.Struct rather than this implementation. We didn't go with anypb in this case because we do want Struct fields to map to JSON Objects, and the Statement cannot dictate the format of the Struct.

Reply:
> Thank you for your comments! I must admit I'm not a very proficient Go programmer, so some of the oddities in my implementation might stem from that.

Hi! No problem, happy to help. And don't worry, I am fairly new with Go too :)

> I think the issue of added fields in messages requiring changes to the function signature is much more likely to occur for predicates, as we don't expect the Statement layer to change very often, especially within the same major version.

That's fair.

> I really like your suggestion to extend the pre-generated message (I tried to find whether this is possible, and couldn't find any info)

There is a way to add validations annotations to your proto fields by using [proto-gen-validate](https://github.com/bufbuild/protoc-gen-validate), once you add the annotations, the generated code will have a `Validate` method. You can find an example of how I use that [here](https://github.com/chainloop-dev/chainloop/blob/ac7d2f5df369a6000185bf17a49f86557b6c8b1c/app/controlplane/api/controlplane/v1/integrations.proto#L56), then this [validation file](https://github.com/chainloop-dev/chainloop/blob/main/app/controlplane/api/controlplane/v1/integrations.pb.validate.go) gets generated.

That said, `proto-gen-validate` mostly focuses on unary per-field validations and in your case (based on the validation code in this PR), you might need something more custom, in which case a custom method is what you might need.

> Don't disagree... what do you suggest as an alternative? This seems like a general problem with using structpb.Struct rather than this implementation.

Indeed, nothing specific about this implementation, my implicit point was that the wrapper was not helping with one of the hardest parts of the construction, the creation of protopb.Structs, specially in the case of annotations. If that was not the case I would have seen a good reason for having this wrapper. In the future, the wrappers (wherever the are) could maybe accept a subset of `map[x]interface` and then internally it will generate the structpb, check the error and so on, simplifying the call.  

> We didn't go with anypb in this case because we do want Struct fields to map to JSON Objects, and the Statement cannot dictate the format of the Struct.

re alternative: honestly I don't know, you have a convincing point on the JSON compatibility. Now I am curious, I'll do some research :)

Thanks for this, looking good! Can't wait to adopt it :)

Reply:
@migmartri Made some updates addressing several of your comments. Hopefully that works!

> There is a way to add validations annotations to your proto fields by using [proto-gen-validate](https://github.com/bufbuild/protoc-gen-validate), once you add the annotations, the generated code will have a Validate method. You can find an example of how I use that [here](https://github.com/chainloop-dev/chainloop/blob/ac7d2f5df369a6000185bf17a49f86557b6c8b1c/app/controlplane/api/controlplane/v1/integrations.proto#L56), then this [validation file](https://github.com/chainloop-dev/chainloop/blob/main/app/controlplane/api/controlplane/v1/integrations.pb.validate.go) gets generated.

Learn something new every day :) Very cool, thanks!

> If that was not the case I would have seen a good reason for having this wrapper. In the future, the wrappers (wherever the are) could maybe accept a subset of map[x]interface and then internally it will generate the structpb, check the error and so on, simplifying the call.

Really good point. Yeah, I imagined this is the sort of thing you had in mind given your previous comment. In many cases, we imagine that the protobufs will actually be populated based on some serialized data, so I wonder about the broader utility of a wrapper that populates the proto messages directly, given libraries like `protojson` and such.